torch/cuda/amp/grad_scaler.py

```diff
@@ -262,26 +264,45 @@ def _unscale_grads_(
                         # For scaled fp16 values, there's a good chance coalescing will cause overflow,
                         # so we should check the coalesced _values().
                         if param.grad.dtype is torch.float16:
-                            param.grad = param.grad.coalesce()
```

Comment:
amp/GradScaler call `coalesce` on fp16 tensors directly
FSDP ShardedGradScaler do fp16->fp32->coalesce->fp16 for numerics

Reply:
What accumulation dtype is used for the addition in `coalesce()`? If it is already accumulating sums in fp32, then I think directly calling `coalesce()` on the fp16 uncoalesced tensor is equivalent to what upcast here.

More broadly, I am curious why `ShardedGradScaler` should have different numeric behavior here compared to `GradScaler`?